identity/v2/service.go

```diff
@@ -0,0 +1,425 @@
+	if !ok {
+		return nil
+	}
+
+	blockHash, err := s.centAPI.GetBlockHash(uint64(revokedAt))
+
+	if err != nil {
+		s.log.Errorf("Couldn't retrieve block hash: %s", err)
+
+		return ErrBlockHashRetrieval
+	}
+
+	block, err := s.centAPI.GetBlock(blockHash)
+
+	if err != nil {
+		s.log.Errorf("Couldn't retrieve block: %s", err)
+
+		return ErrBlockRetrieval
+	}
+
+	meta, err := s.centAPI.GetMetadataLatest()
+
+	if err != nil {
+		s.log.Errorf("Couldn't retrieve metadata: %s", err)
+
+		return podErrors.ErrMetadataRetrieval
+	}
+
+	timestamp, err := getBlockTimestamp(meta, block)
+
+	if err != nil {
+		s.log.Errorf("Couldn't retrieve metadata: %s", err)
+
+		return ErrBlockTimestampRetrieval
+	}
+
+	if validationTime.After(*timestamp) {
+		s.log.Error("Key is revoked")
+
+		return ErrKeyRevoked
+	}
+
+	return nil
+}
+
+func (s *service) ValidateSignature(
+	ctx context.Context,
+	accountID *types.AccountID,
+	pubKey []byte,
+	message []byte,
+	signature []byte,
+	validationTime time.Time,
+) error {
+	if err := s.ValidateKey(ctx, accountID, pubKey, keystoreType.KeyPurposeP2PDocumentSigning, validationTime); err != nil {
+		s.log.Errorf("Couldn't validate key: %s", err)
+
+		return err
+	}
+
+	if !crypto.VerifyMessage(pubKey, message, signature, crypto.CurveEd25519) {
+		s.log.Error("Couldn't verify message - invalid signature")
+
+		return ErrInvalidSignature
+	}
+
+	return nil
+}
+
+func (s *service) ValidateAccount(ctx context.Context, accountID *types.AccountID) error {
+	if err := validation.Validate(validation.NewValidator(accountID, accountIDValidatorFn)); err != nil {
+		s.log.Errorf("Invalid account ID: %s", err)
+
+		return err
+	}
+
+	meta, err := s.centAPI.GetMetadataLatest()
+
+	if err != nil {
+		s.log.Errorf("Couldn't get latest metadata: %s", err)
+
+		return ErrMetadataRetrieval
+	}
+
+	accountStorageKey, err := types.CreateStorageKey(meta, "System", "Account", accountID.ToBytes())
+
+	if err != nil {
+		s.log.Errorf("Couldn't create storage key for account: %s", err)
+
+		return ErrAccountStorageKeyCreation
+	}
+
+	var accountInfo types.AccountInfo
+
+	ok, err := s.centAPI.GetStorageLatest(accountStorageKey, &accountInfo)
+
+	if err != nil {
+		s.log.Errorf("Couldn't retrieve account from storage: %s", err)
+
+		return ErrAccountStorageRetrieval
+	}
+
+	if ok {
+		return nil
+	}
+
+	// Account info not found, check if account is a valid anonymous proxy i.e. has at least one proxy with type Any.
+	return s.isValidAnonymousProxy(ctx, accountID)
+}
+
+func (s *service) isValidAnonymousProxy(ctx context.Context, accountID *types.AccountID) error {
+	res, err := s.proxyAPI.GetProxies(ctx, accountID)
+
+	if err != nil {
+		s.log.Errorf("Couldn't retrieve account proxies: %s", err)
+
+		if errors.Is(err, proxy.ErrProxiesNotFound) {
+			return ErrAccountNotAnonymousProxy
+		}
+
+		return ErrAccountProxiesRetrieval
+	}
+
+	for _, proxyDefinition := range res.ProxyDefinitions {
+		if uint8(proxyDefinition.ProxyType) == uint8(proxyType.Any) {
```

Comment:
I dont think we can enforce that every proxy has at least one delegate that has type Any. I think is enough that it returns a non-empty list of proxyDefinitions